client/cli/src/runtime.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures::{Future, future, future::FutureExt};
+use futures::select;
+use futures::pin_mut;
+use sc_service::{AbstractService, Configuration};
+use crate::error;
+use crate::informant;
+
+async fn main<F, E>(func: F) -> Result<(), Box<dyn std::error::Error>>
+where
+	F: Future<Output = Result<(), E>> + future::FusedFuture + Unpin,
+	E: 'static + std::error::Error,
+{
+	use tokio::signal::unix::{signal, SignalKind};
+
+	let mut stream_int = signal(SignalKind::interrupt())?;
+	let mut stream_term = signal(SignalKind::terminate())?;
+
+	let t1 = stream_int.recv().fuse();
+	let t2 = stream_term.recv().fuse();
+	let mut t3 = func;
+
+	pin_mut!(t1, t2);
+
+	select! {
+		_ = t1 => println!("Caught SIGINT"),
+		_ = t2 => println!("Caught SIGTERM"),
+		res = t3 => res?,
+	}
+
+	Ok(())
+}
+
+fn build_runtime() -> Result<tokio::runtime::Runtime, std::io::Error> {
+	tokio::runtime::Builder::new()
+		.thread_name("main-tokio-")
+		.threaded_scheduler()
+		.enable_all()
+		.build()
+}
+
+/// A helper function that runs a future with tokio and stops if the process receives the signal
+/// SIGTERM or SIGINT
+pub fn run_until_exit<FUT, ERR, G, E, F>(
+	mut config: Configuration<G, E>,
+	future_builder: F,
+) -> error::Result<()>
+where
+	F: FnOnce(Configuration<G, E>) -> error::Result<FUT>,
+	FUT: Future<Output = Result<(), ERR>> + future::Future,
+	ERR: 'static + std::error::Error,
+{
+	let mut runtime = build_runtime()?;
+
+	config.tasks_executor = {
+		let runtime_handle = runtime.handle().clone();
+		Some(Box::new(move |fut| { runtime_handle.spawn(fut); }))
+	};
+
+	let f = future_builder(config)?;
+	let f = f.fuse();
+	pin_mut!(f);
+
+	runtime.block_on(main(f)).map_err(|e| e.to_string())?;
+
+	Ok(())
+}
+
+async fn run_service_with_informant<T>(service: T) -> error::Result<()>
+where
+	T: AbstractService + Unpin,
+{
+	let informant_future = informant::build(&service);
+	let handle = tokio::spawn(informant_future);
```

Comment:
This is okay-ish because this entire module uses tokio. Ideally I'd prefer to pass an abstraction over an executor by parameter to this function.

Reply:
Actually you can set the `task_executor` yourself and don't use `run_until_exit` at all. That's what is happening with WASM.

Reply:
What I mean is that `run_service_with_informant` is here an implementation detail of `run_service_until_exit`, and `run_service_until_exit` uses tokio, so it's ok that `run_service_with_informant` relies on tokio being there.

However, if someone is to mark `run_service_with_informant` as `pub`, then we would have to pass an executor as parameter.

My reasoning is basically: we could do either this now in order to avoid future accidents, or not care.

Reply:
I'm fixing that now but I just realized `tasks_executor` is not proper English. Correct me if I'm wrong but I think it's `task_executor`. Maybe I should fix that now so we're done with it.